Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+  VerifyNotShutdown();
+  auto shared_self = shared_from_this();
+  worker_queue()->Enqueue([shared_self, callback] {
+    shared_self->remote_store_->EnableNetwork();
+    if (callback) {
+      shared_self->user_executor()->Execute([=] { callback(Status::OK()); });
+    }
+  });
+}
+
+void FirestoreClient::Shutdown(StatusCallback callback) {
+  auto shared_self = shared_from_this();
+  worker_queue()->EnqueueAndInitiateShutdown([shared_self, callback] {
+    shared_self->credentials_provider_->SetCredentialChangeListener(nullptr);
+
+    // If we've scheduled LRU garbage collection, cancel it.
+    if (shared_self->lru_callback_) {
+      shared_self->lru_callback_.Cancel();
+    }
+    shared_self->remote_store_->Shutdown();
+    [shared_self->persistence_ shutdown];
+  });
+
+  // This separate enqueue ensures if shutdown is called multiple times
+  // every time the callback is triggered. If it is in the above
+  // enqueue, it might not get executed because after first shutdown
+  // all operations are not executed.
+  worker_queue()->EnqueueEvenAfterShutdown([shared_self, callback] {
+    if (callback) {
+      shared_self->user_executor()->Execute([=] { callback(Status::OK()); });
+    }
+  });
+}
+
+void FirestoreClient::WaitForPendingWrites(StatusCallback callback) {
+  VerifyNotShutdown();
+
+  // Dispatch the result back onto the user dispatch queue.
+  auto shared_self = shared_from_this();
+  auto async_callback = [shared_self, callback](util::Status status) {
+    if (callback) {
+      shared_self->user_executor()->Execute(
+          [=] { callback(std::move(status)); });
+    }
+  };
+
+  worker_queue()->Enqueue([shared_self, async_callback] {
+    [shared_self->sync_engine_
+        registerPendingWritesCallback:std::move(async_callback)];
+  });
+}
+
+void FirestoreClient::VerifyNotShutdown() {
+  if (is_shutdown()) {
+    ThrowIllegalState("The client has already been shutdown.");
+  }
+}
+
+bool FirestoreClient::is_shutdown() const {
+  // Technically, the worker queue is still running, but only accepting tasks
+  // related to shutdown or supposed to be run after shutdown. It is effectively
+  // shut down to the eyes of users.
+  return worker_queue()->is_shutting_down();
+}
+
+std::shared_ptr<QueryListener> FirestoreClient::ListenToQuery(
+    Query query,
+    ListenOptions options,
+    ViewSnapshot::SharedListener&& listener) {
+  VerifyNotShutdown();
+
+  auto query_listener = QueryListener::Create(
+      std::move(query), std::move(options), std::move(listener));
+
+  auto shared_self = shared_from_this();
+  worker_queue()->Enqueue([shared_self, query_listener] {
+    shared_self->event_manager_->AddQueryListener(std::move(query_listener));
+  });
+
+  return query_listener;
+}
+
+void FirestoreClient::RemoveListener(
+    const std::shared_ptr<QueryListener>& listener) {
+  // Checks for shutdown but does not throw error, allowing it to be an no-op if
+  // client is already shutdown.
+  if (is_shutdown()) {
```

Comment:
You could make this a one liner if you wanted.

Reply:
done.